### PR TITLE
Feature/cdap 14644 checkpointing for gcs with default dir

### PR DESCRIPTION
**Pull request by jayfjin**
Add checkpointing for GCS, and use a default directory if checkpointDir is not set. 

Builds:
https://builds.cask.co/browse/CDAP-DUT6918-1

### CODE REVIEW

**Review by sreevatsanraman — APPROVED** [review]

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java [hunk 1]
```diff
@@ -80,26 +92,74 @@ public void run(final JavaSparkExecutionContext sec) throws Exception {
...
+                              FileSetProperties.builder().build());
         }
-      }, Exception.class);
+        // there isn't any way to instantiate the fileset except in a TxRunnable, so need to use a reference.
+        final AtomicReference<Location> checkpointBaseRef = new AtomicReference<>();
+        Transactionals.execute(sec, new TxRunnable() {
+          @Override
+          public void run(DatasetContext context) throws Exception {
+            FileSet checkpointFileSet = context.getDataset(DEFAULT_CHECKPOINT_DATASET_NAME);
+            checkpointBaseRef.set(checkpointFileSet.getBaseLocation());
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java:112]: Not a biggie but: looks like the dataset is not really used anywhere. You should be able to discard it here.:
```
context.discardDataset(checkpointFileSet)
```
That will close the dataset and remove it from the dataset cache.

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java [hunk 2]
```diff
@@ -80,26 +92,74 @@ public void run(final JavaSparkExecutionContext sec) throws Exception {
...
+
+      context = new JavaSparkContext();
+      Configuration configuration = context.hadoopConfiguration();
+      // Set the filesystem to whatever the checkpoint directory uses. This is necessary since spark will override
+      // the URI schema with what is set in this config. This needs to happen before StreamingCompat.getOrCreate
+      // is called, since StreamingCompat.getOrCreate will attempt to parse the checkpointDir before calling
+      // context function.
+      URI checkpointUri = checkpointDirPath.toUri();
+      if (checkpointUri.getScheme() != null) {
+        configuration.set("fs.defaultFS", checkpointDir);
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java:129]: does this work reliably? You can use a full path as the default file system?

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java [hunk 3]
```diff
@@ -80,26 +92,74 @@ public void run(final JavaSparkExecutionContext sec) throws Exception {
...
-      Transactionals.execute(sec, new TxRunnable() {
-        @Override
-        public void run(DatasetContext context) throws Exception {
-          FileSet checkpointFileSet = context.getDataset(DataStreamsApp.CHECKPOINT_FILESET);
-          checkpointBaseRef.set(checkpointFileSet.getBaseLocation());
+      String configCheckpointDir = pipelineSpec.getCheckpointDirectory();
+      if (Strings.isNullOrEmpty(configCheckpointDir)) {
+        // Use the directory of a fileset dataset if the checkpoint directory is not set.
+        Admin admin = sec.getAdmin();
+        if (!admin.datasetExists(DEFAULT_CHECKPOINT_DATASET_NAME)) {
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java:102]: I guess this means that all Spark pipelines in a namespace share, by default, the same checkpoint directory (this dataset's base path). Will this work in the future in multi-tenant setups with more fine grained access control?

**Review by anew — APPROVED** [review]
I had a few minor comments, actually questions. LGTM

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java (hunk 2 above)
Comment (jayfjin) [reply cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java:129]: yes

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java (hunk 3 above)
Comment (jayfjin) [reply cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java:102]: We may need to prefix the path in the future with an additional field

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java (hunk 1 above)
Comment (jayfjin) [reply cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java:112]: Gotcha, sorry I merged after seeing the build passed before the page refreshed with your comments. Will send a followup

**Pull request merged**